### DISCUSSION TITLE
Porting to Rust

### DISCUSSION THREAD

**Discussion by mmlngl** [donmccurdy/glTF-Transform#714] [Ideas]
Hi there 👋

Huge fan of this lib and all the hard work done.

Out of respect for the project, I thought I'd try and gradually port it to rust under the same licence.

Any tips on where to start? I was thinking `inspect` or something?

Thanks

**Comment by donmccurdy** [comment]
Hi @mmmoli! That sounds like a great project, no objections from me! I'm not a lawyer, but as far as I'm concerned you're welcome to choose whatever license you think is appropriate for the port.

As far as where to start, I made two fairly important decisions early in this project:

1. Users of the library would be able to define their own [extensions](https://gltf-transform.donmccurdy.com/extensions.html)
2. Most of the long-term time and effort would be spent implementing various _operations on glTF data_, and so spending time up front making it easier/faster to write those operations would pay off

Lots of other decisions stem from these. With that said —

You'll want to decide before implementing too many operations (like `inspect()`, `draco()`, etc.) what the "things" are that functions operate on. The [Concepts](https://gltf-transform.donmccurdy.com/concepts.html) page explains which glTF concepts are exposed in this library's public API, and which are internalized. Those concepts are implemented by a [property graph](https://github.com/donmccurdy/property-graph). When this code runs ...

```javascript
import { listTextureSlots } from '@gltf-transform/functions';

const slots = listTextureSlots(document, texture);
// → ['occlusionTexture', 'metallicRoughnessTexure']
```

... it is searching up the graph for materials and material extensions holding references to the given texture. There are a lot of similar operations in the library. I'm not sure what an Rust-friendly implementation of a property graph would look like — it might be a bit different from my implementation, which caters to TypeScript's type checking system.

If you look at any of the data classes ([material.ts](https://github.com/donmccurdy/glTF-Transform/blob/ce526e7dfcd0c0db361f0d7771e687478fccaf1e/packages/core/src/properties/material.ts#L89-L97) is a good example), each is defining public getters/setters that internally call the property graph APIs. I'd suggest starting by having a look at those decisions, and adapting them to Rust and to your ideas however makes the most sense to you. Best of luck and feel free to post questions whenever. 🙂
